MDEV-27143 Different result when group by with correct function

MarcoZhang111 opened the pull request:
https://jira.mariadb.org/browse/MDEV-27143

<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling this template <3

If you have any questions related to MariaDB or you just want to
hang out and meet other community members, please join us on
https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue
that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-27143*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed, what was it looking like before
   the change and how it's looking with this patch applied
3. Do you think this patch might introduce side-effects in
   other parts of the server?
-->
## Description
TODO: fill description here

## How can this PR be tested?

TODO: modify the automated test suite to verify that the PR causes MariaDB to
behave as intended. Consult the documentation on
["Writing good test cases"](https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server).
In many cases, this will be as simple as modifying one `.test` and one `.result`
file in the `mysql-test/` subdirectory. Without _automated_ tests, future regressions
in the expected behavior can't be automatically detected and verified.

If the changes are not amenable to automated testing, please explain why not and
carefully describe how to test manually.

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
-->
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

<!--
You might consider answering some questions like:
1. Does this affect the on-disk format used by MariaDB?
2. Does this change any behavior experienced by a user
   who upgrades from a version prior to this patch?
3. Would a user be able to start MariaDB on a datadir
   created prior to your fix?
-->
## Backward compatibility
TODO: fill details here, if applicable, or remove the section

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=1955) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=1955) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=1955) it.</sub>

grooverdan commented:
@MarcoZhang111 first of all, thank you very much for this fix.

Being a fix you can apply the fix against the 10.2 branch (is it the same fix)?

Can you include more information in your commit messages as to why this is the correct fix?

Are you able to include your test case in mysql-test/t/olap.test, and the `mtr --record olap.test` to generate the result file, check its right, and the commit those too?

MarcoZhang111 commented:
Hi，grooverdan thanks for your reply. 
I just want to eliminate the influence when the group by list with some functions，for examaple in my MDEV-27413 the sql command select c, count(*) from t group by c desc with rollup  and select c, count(*) from t group by convert(c, date) desc with rollup  should be have the same result ，but it haven't.

grooverdan commented:
Ok. adjust title when you ready.

grooverdan commented:
@MarcoZhang111 do you need a hand resolving the merge conflict, rebasing to 10.2 and adding the test case from the JIRA issue? I can look at doing that for you next week if you are happy the code change is all there.

MarcoZhang111 committed:
MDEV-27143 Different result when group by with correct function

By default, a typical ROLLUP with a non-aggregated column (c) will
have NULL as its value in the last line, the ROLLUP row.

create table t (c int);
insert into t values (20121212),(9),(19800101),(20121212),(19800101);

select c, count(*) from t group by  c desc with rollup;

MariaDB [test]> select c, count(*) from t group by c desc with rollup;
+----------+----------+
| c        | count(*) |
+----------+----------+
| 20121212 |        2 |
| 19800101 |        2 |
|        9 |        1 |
|     NULL |        5 |
+----------+----------+

When however, a function is used in GROUP BY, MariaDB incorrectly uses
the last row as the column value. Like the previous case, this should
be NULL.

select c, count(*) from t group by convert(c,int) desc with rollup;
+----------+----------+
| c        | count(*) |
+----------+----------+
| 20121212 |        2 |
| 19800101 |        2 |
|        9 |        1 |
|        9 |        5 |
+----------+----------+

MarcoZhang111 commented:
Thanks grooverdan ,I have created a new commit in branch 10.2 for this issue ,need i creat a new pull request for this commit ?

grooverdan commented:
Edit this pull request at the top and change the destination branch to 10.2
Then you can push your 10.2 commit to remote branch called 10.8 and it will show up here. Don't worry about the name, it can't be changed.

We'll merge the change from 10.2 up to 10.8.

MarcoZhang111 commented:
/recheck

grooverdan commented:
> 

You still need to force push your 10.2 branch containing the fix here

Assuming your commit(s) are just on the 10.2 branch:
```
git push -f 10.2:10.8
```

MarcoZhang111 commented:
Hi，grooverdan
  I have signed the cla ，but the check have not passed，do you konw why？

grooverdan commented:
The cla check wasn't passing because of the other commits where other authors were not signed with the CLA github tool.

I corrected this by rebasing the commit removing the merge commit that you had.

While I did this I also:
* A small restructured your code for read ability
* Added the test case from the MDEV.
* Added some wording in the commit message.

Can you see if my commit message wording makes sense?

I tested `select c, count() from t group by convert(c, date) desc with rollup`
```
+select c, count(*) from t group by convert(c, date) desc with rollup;
+c	count(*)
+20121212	2
+19800101	2
+9	1
+NULL	5
+Warnings:
+Warning	1292	Incorrect datetime value: '9' for column `test`.`t`.`c` at row 1
+Warning	1292	Incorrect datetime value: '9' for column `test`.`t`.`c` at row 5
+Warning	1292	Incorrect datetime value: '9' for column `test`.`t`.`c` at row 5
```
And it seems to work, though the warnings are odd.

What I notice isn't working is:
```+select c, count(*) from t group by c+1 desc with rollup;
+c	count(*)
+20121212	2
+19800101	2
+9	1
+9	5
```

Maybe the condition before your code change, `(*group_tmp->item == item)` just needs to be a test if `item` is not an aggregate function?

MarcoZhang111 commented:
Thanks for your responding ,grooverdan.
  I can make sense your commit. I can promise if we group by with  correct function   ，the reasults have no warnings and  will be the same  with without function，in your test，the table structure is very simple ,so your test was not different .
I tested with this :
Group by without functiong:
MariaDB [test]> select country_id,product ,year ,sum(profit) from t1 group by country_id,year,product with rollup;
+------------+------------+------+-------------+
| country_id | product    | year | sum(profit) |
+------------+------------+------+-------------+
|          1 | Calculator | 1999 |          50 |
|          1 | Computer   | 1999 |        2700 |
|          1 | TV         | 1999 |         250 |
|          1 | NULL       | 1999 |        3000 |
|          1 | Calculator | 2000 |          75 |
|          1 | Computer   | 2000 |        1500 |
|          1 | TV         | 2000 |         150 |
|          1 | NULL       | 2000 |        1725 |
|          1 | NULL       | NULL |        4725 |
|          2 | Calculator | 2000 |         150 |
|          2 | Computer   | 2000 |        2700 |
|          2 | TV         | 2000 |         200 |
|          2 | NULL       | 2000 |        3050 |
|          2 | NULL       | NULL |        3050 |
|          3 | Phone      | 2003 |          10 |
|          3 | NULL       | 2003 |          10 |
|          3 | NULL       | NULL |          10 |
|       NULL | NULL       | NULL |        7785 |
+------------+------------+------+-------------+
18 rows in set (0.000 sec)
Groub by with correct function:
MariaDB [test]> select country_id,product ,year ,sum(profit) from t1 group by country_id,convert(year,char),product with rollup;
+------------+------------+------+-------------+
| country_id | product    | year | sum(profit) |
+------------+------------+------+-------------+
|          1 | Calculator | 1999 |          50 |
|          1 | Computer   | 1999 |        2700 |
|          1 | TV         | 1999 |         250 |
|          1 | NULL       | 1999 |        3000 |
|          1 | Calculator | 2000 |          75 |
|          1 | Computer   | 2000 |        1500 |
|          1 | TV         | 2000 |         150 |
|          1 | NULL       | 2000 |        1725 |
|          1 | NULL       | NULL |        4725 |
|          2 | Calculator | 2000 |         150 |
|          2 | Computer   | 2000 |        2700 |
|          2 | TV         | 2000 |         200 |
|          2 | NULL       | 2000 |        3050 |
|          2 | NULL       | NULL |        3050 |
|          3 | Phone      | 2003 |          10 |
|          3 | NULL       | 2003 |          10 |
|          3 | NULL       | NULL |          10 |
|       NULL | NULL       | NULL |        7785 |
+------------+------------+------+-------------+
18 rows in set (0.000 sec)
Groub by with warning function:MariaDB [test]> select country_id,product ,year ,sum(profit) from t1 group by country_id,convert(year,date),product with rollup;
+------------+------------+------+-------------+
| country_id | product    | year | sum(profit) |
+------------+------------+------+-------------+
|          1 | Calculator | 1999 |         125 |
|          1 | Computer   | 1999 |        4200 |
|          1 | TV         | 1999 |         400 |
|          1 | NULL       | 1999 |        4725 |
|          1 | NULL       | NULL |        4725 |
|          2 | Calculator | 2000 |         150 |
|          2 | Computer   | 2000 |        2700 |
|          2 | TV         | 2000 |         200 |
|          2 | NULL       | 2000 |        3050 |
|          2 | NULL       | NULL |        3050 |
|          3 | Phone      | 2003 |          10 |
|          3 | NULL       | 2003 |          10 |
|          3 | NULL       | NULL |          10 |
|       NULL | NULL       | NULL |        7785 |
+------------+------------+------+-------------+
14 rows in set, 37 warnings (0.000 sec)

You can try to test like this.

grooverdan commented:
Right. Looking at example data from https://mariadb.com/kb/en/select-with-rollup/, the following query skips all `ROLLUP` values.

```
MariaDB [test]> SELECT country, year, genre, SUM(sales)    FROM booksales GROUP BY concat('a','b',length(country)), year+1, length(genre)+1 WITH ROLLUP;
```

Should `(*group_tmp->item == item)` some how be that `*group_tmp` recursively contains `item`? What about `group by` that contains multiple result fields in a function? What [do other implementations](https://dbfiddle.uk) do?

MarcoZhang111 commented:
In my opinion,whatever result group by contains in a function，the group_tmp with the item_ptr should be point a item field ,right ? we can search the item filed and set this item null.

grooverdan commented:
> In my opinion,whatever result group by contains in a function，
> the group_tmp with the item_ptr should be point a item field ,right ? we can search the item filed and set this item null.

It will be an item, but it may not the be the simple item or the first item. See the examples in my previous query and expand the test case to include a wide variety of these. Its also not necessary for the `GROUP BY` to be in the same order as the `SELECT`. A confirmation by set of test cases considered would be appreciated. What do you expect with a `GROUP BY f(a,b) WITH ROLLUP`? This might need something like `Item_func::traverse_cond` or `Item::walk`.

MarcoZhang111 commented:
Can we find the filed item through the group_tmp->item_ptr ? I found the group_tmp->item_ptr is the pointer to the group by function and this tail pointer of the list should be point the item field . So, i think we can use the function item->eq(group_tmp->item_ptr(the tail pointer),0) as the judgment condition to set the item null. No matter what function is followed by group by, the rollup rule should not change.

gkodinov commented:
According to the accompanying Jira, this has been fixed and closed by [MDEV-19039](https://jira.mariadb.org/browse/MDEV-19039). Closing the PR too. Please submit a new pull request (and eventually a new Jira if the problem is different) if you are still seeing it.
